### PR TITLE
Add transactionAmount field with criteria search

### PR DESCRIPTION
**Pull request by mnovelodou**
## What

Adds a `transactionAmount` field to expenses that records the amount of the original (pre-split) transaction a line was derived from, plus an exact-match search over it.

When one real-world transaction is split into multiple expense lines (e.g. a \$100 purchase → Groceries \$60 + Household \$40), each line can carry the original \$100 total so all lines from that transaction can later be found together.

## Changes

- **Schema**: additive nullable `transaction_amount NUMERIC(15,2)` column. **No index** in this change (deferred by request).
- **Field plumbing**: added to the `Expense` DTO, `ExpenseEntity`, mapper, and the repository INSERT/bulk/UPDATE/RowMapper.
- **Create / bulk-create**: when the caller omits `transactionAmount`, it defaults to the line's own `amount`. Never derived from splits.
- **Update is partial** for `amount` and `transactionAmount`: omitting either preserves the stored value. A null `amount` on update no longer returns 400, and a null `transactionAmount` is **not** re-defaulted to `amount`.
- **Search**: `GET /expenses?transaction_amount=` adds an exact-match filter that reuses the existing mandatory `created_by` + date-range criteria search and the existing cursor pagination. Combinable with `category`/`subcategory`/`account_id`; not encoded in the cursor (resupply per page).

## Notes for reviewers

- The filter is exposed only on `GET /expenses`; `listByAccount` passes `null` (signature unchanged), so account-scoped listing has no `transaction_amount` filter by design.
- The default-to-`amount` rule applies **only at creation**, never on update — update changes a field only when explicitly provided.
- Existing clients are unaffected: the field and filter are optional; pre-existing rows keep `transaction_amount = null` (no backfill).
- Planned via OpenSpec change `add-transaction-amount` (proposal/design/specs/tasks included under `openspec/changes/`).

## Testing

`./gradlew test integrationTest` — green (216 unit + 107 integration). New coverage: repository filter binding, service-layer create defaulting + partial-update preservation, controller param binding, and an end-to-end split-and-search integration test with cursor pagination.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: openspec/changes/add-transaction-amount/specs/list-expenses-by-user/spec.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+## MODIFIED Requirements
+
+### Requirement: List expenses by user with date window
+The system SHALL expose a `GET /expenses` endpoint that returns a paginated list of expenses for a given user. `user_id` SHALL be a required query parameter. Once authentication is introduced, `user_id` will be inferred from the OAuth token and this parameter will be removed. `start_date` and `end_date` are both optional and independently defaulted:
```
Comment (mnovelodou) [inline openspec/changes/add-transaction-amount/specs/list-expenses-by-user/spec.md:4]: CODEX: This requirement says `user_id` is mandatory and the missing-parameter scenario expects HTTP 400, but `ExpenseController.list` still declares `user_id` with `required = false`, and `ExpenseService.listByUser` defaults a missing/blank value to the authenticated caller. Please either enforce the required parameter in this change or update this spec so the documented contract matches the implemented behavior.

File: openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+## ADDED Requirements
+
+### Requirement: Expense records an optional original transaction amount
+An expense SHALL carry an optional `transactionAmount` field (decimal) representing the amount of the original real-world transaction it was derived from. The value SHALL be supplied by the API caller only; the system SHALL NOT derive `transactionAmount` from `amount` changes or split logic. When the caller omits `transactionAmount` on a write, the system SHALL default it to that line's `amount` before persisting, and the resolved value SHALL be returned in the response. When a single transaction is split into multiple expense lines, each line SHALL be able to carry the same explicitly-supplied `transactionAmount` equal to the original transaction total, independent of each line's own `amount`. The field SHALL be accepted on create (`POST /expenses`), bulk create (`POST /expenses/bulk`), and update (`PUT /expenses/{id}`), and SHALL be returned on every expense representation.
```
Comment (mnovelodou) [inline openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md]: CODEX: The sentence “When the caller omits `transactionAmount` on a write, the system SHALL default it” also covers `PUT`, which contradicts the update requirement below that omission preserves the stored value. The code follows the latter behavior, but this top-level requirement should explicitly say create/bulk-create only; otherwise a future implementation could legitimately reintroduce defaulting on update.

File: src/main/java/com/novelosoftware/expenses/dto/Expense.java [hunk 3]
```diff
@@ -15,6 +15,8 @@ public record Expense(
     Long accountId,
     /** amount of the expense */
     BigDecimal amount,
+    /** amount of the original (pre-split) transaction this line was derived from */
+    BigDecimal transactionAmount,
     /** description of the expense */
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/dto/Expense.java:20]: CODEX: `transactionAmount` accepts an unrestricted `BigDecimal`, while persistence is `NUMERIC(15,2)`. Values with more than 13 integer digits will reach PostgreSQL and fail as a server error, and values with extra fractional digits may be rounded by the database, so the returned value can differ from the request. Please validate precision/scale at the API/service boundary (and return a 400) before create, bulk-create, or update.

**Review by mnovelodou — COMMENTED** [review]
CODEX: Review focused on consistency with the OpenSpec change, maintainability, edge cases, and scalability. I did not run tests, per scope.

File: openspec/changes/add-transaction-amount/specs/list-expenses-by-user/spec.md (hunk 1 above)
Comment (mnovelodou) [reply openspec/changes/add-transaction-amount/specs/list-expenses-by-user/spec.md:4]: Good catch, but we're intentionally not addressing this in this PR. The mismatch is pre-existing: owner-scoped-authorization already made `user_id` optional (`required = false`) and `listByUser` default a blank value to the authenticated caller (cross-user hidden as 404). The stale "required → 400" text comes from the original `list-expenses-by-user` main spec, and I copied the requirement block verbatim into this delta. I believe a prior change updated the behavior without running archive/sync to bring the main spec in line, so the source of truth is out of date — not something this PR should patch. I'll reconcile the main spec separately via the OpenSpec sync/archive flow so this delta isn't carrying an unrelated contract fix.

File: openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md (hunk 2 above)
Comment (mnovelodou) [reply openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md]: Fixed in 3e0f... — the top-level requirement now scopes the "default to `amount`" rule explicitly to create (`POST /expenses`) and bulk create (`POST /expenses/bulk`), and states the defaulting does NOT apply on update, pointing to the "Update preserves omitted amount and transaction amount" requirement. So a future implementation can't legitimately reintroduce defaulting on PUT.

File: src/main/java/com/novelosoftware/expenses/dto/Expense.java (hunk 3 above)
Comment (mnovelodou) [reply src/main/java/com/novelosoftware/expenses/dto/Expense.java:20]: Fixed for both `amount` and `transactionAmount` (the same gap existed on `amount` already). `ExpenseService` now validates monetary values against the NUMERIC(15,2) bounds in `expenseWriteValidations` on create/bulk/update: a value with more than 2 fractional digits or more than 13 integer digits is rejected with HTTP 400, instead of letting PostgreSQL silently round it or fail with a server error. Null is skipped (it's defaulted on create or preserved on update). Added unit tests (too-many-decimals and too-many-integer-digits on both create and update) plus an end-to-end IT, and a "Monetary values bounded to NUMERIC(15,2)" requirement with scenarios in the spec.

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java [hunk 4]
```diff
@@ -141,7 +144,9 @@ public UpdateExpenseResponse update(Long id, UpdateExpenseRequest request) {
             throw createExpenseNotFoundException(id);
         }
 
-        ExpenseEntity newExpense = ExpenseMapper.toEntity(expense);
+        // Preserve the stored amount / transactionAmount when the request omits them. The
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/services/ExpenseService.java:153]: CODEX: This read/merge/full-update sequence can violate the new partial-update semantics under concurrency. For example, request A changes only `amount` and request B changes only `transactionAmount`; if both read the same old row, whichever writes last also writes its stale copy of the field it omitted, losing the other request’s change. Consider preserving omitted fields atomically in the SQL update (for example with `COALESCE`/dynamic SET clauses), or add optimistic locking so concurrent updates fail instead of silently overwriting each other.

File: openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+- **THEN** the system returns HTTP 400 and no expense is created
+
+#### Scenario: More than thirteen integer digits rejected
+- **WHEN** a `POST /expenses` request supplies `amount = 12345678901234` (14 integer digits)
+- **THEN** the system returns HTTP 400 and no expense is created
+
+### Requirement: Search expenses by transaction amount
+The `GET /expenses` criteria search SHALL accept an optional `transaction_amount` query parameter. When supplied, results SHALL be restricted to expenses whose `transaction_amount` equals the supplied value exactly. When absent, `transaction_amount` SHALL NOT constrain results. The filter reuses the existing mandatory `created_by` + date-range criteria search and the existing forward cursor pagination; the `transaction_amount` filter SHALL NOT be encoded in the cursor and the caller SHALL supply it on every page request. `transaction_amount` MAY be combined with the `category`, `subcategory`, and `accountId` filters. This change SHALL NOT add a database index for the new column.
+
+#### Scenario: Filter by transaction amount returns matching split lines
```
Comment (mnovelodou) [inline openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md:51]: CODEX: `transactionAmount` is not a unique transaction identifier, so this search can return unrelated purchases that happen to have the same total within the selected window. That means it cannot reliably “find all lines from one transaction” as described in the proposal. If exact grouping is a requirement, consider storing a transaction/group ID alongside the amount; otherwise document that the filter is intentionally amount-based and may include false-positive matches.

**Review by mnovelodou — COMMENTED** [review]
CODEX: Second pass completed after reading all replies and reviewing the latest head. The earlier defaulting and numeric-bound concerns are addressed; the pre-existing `user_id` mismatch is acknowledged for separate reconciliation. Two additional risks remain.

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java (hunk 4 above)
Comment (mnovelodou) [reply src/main/java/com/novelosoftware/expenses/services/ExpenseService.java:153]: CODEX: Agreed. Given the current single-user usage and the expectation that users operate on separate records, concurrent writes to the same expense are not a practical concern for this stage. Optimistic locking or atomic patch SQL can remain future hardening; this is non-blocking.

File: openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md (hunk 5 above)
Comment (mnovelodou) [reply openspec/changes/add-transaction-amount/specs/transaction-amount/spec.md:51]: CODEX: Agreed. The field is intentionally an amount-based search key, not a unique transaction identifier; duplicate totals are valid and expected. Results are therefore all matching amounts within the user/date filters, not guaranteed membership in one transaction. This is non-blocking.

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java [hunk 6]
```diff
@@ -263,7 +272,41 @@ private CursorPageResponse<Expense> listForOwner(String userId,
...
+    }
+
+    /**
+     * Validates an expense write payload.
+     *
+     * @param amountRequired when {@code true} a null {@code amount} is rejected (create/bulk);
+     *                       when {@code false} a null {@code amount} is allowed (update treats it
+     *                       as "leave unchanged")
+     */
+    private void expenseWriteValidations(Expense expense, String callerSub, boolean amountRequired) {
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/services/ExpenseService.java]: Don't add booleans here, instead, check the amount presence only when it is required.

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java [hunk 7]
```diff
@@ -297,4 +345,26 @@ private void expenseWriteValidations(Expense expense, String callerSub) {
         // someone else's account are both hidden as 404 (no existence disclosure).
         accountService.getById(expense.accountId(), false);
     }
+
+    /** Monetary columns are NUMERIC(15,2): up to 13 integer digits and 2 fractional digits. */
+    private static final int MONETARY_SCALE = 2;
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/services/ExpenseService.java]: Move this at the top before the class fields.

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java (hunk 6 above)
Comment (mnovelodou) [reply src/main/java/com/novelosoftware/expenses/services/ExpenseService.java]: Done in d6a1e1d. Dropped the `amountRequired` boolean from `expenseWriteValidations`; amount presence is now checked by a small `requireAmount(expense)` that only create and bulkCreate call. Update doesn't call it, so a null amount keeps preserving the stored value.

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java (hunk 7 above)
Comment (mnovelodou) [reply src/main/java/com/novelosoftware/expenses/services/ExpenseService.java]: Done in d6a1e1d. Moved `MONETARY_SCALE` and `MONETARY_INTEGER_DIGITS` to the top of the class, above the instance fields.
